plugin/diagnostics.py

```diff
@@ -292,16 +223,175 @@ def saved(self) -> None:
         ...
 
 
+class DiagnosticsUpdateWalk(object):
+
+    def begin(self) -> None:
+        pass
+
+    def begin_file(self, file_path: str) -> None:
+        pass
+
+    def diagnostic(self, diagnostic: 'Diagnostic') -> None:
+        pass
+
+    def end_file(self, file_path: str) -> None:
+        pass
+
+    def end(self) -> None:
+        pass
+
+
+class DiagnosticViewRegions(DiagnosticsUpdateWalk):
+
+    def __init__(self, view: sublime.View) -> None:
+        self._view = view
+        self._regions = {}  # type: Dict[int, List[sublime.Region]]
+        self._relevant_file = False
+
+    def begin(self) -> None:
+        for severity in self._regions:
+            self._regions[severity] = []
+
+    def begin_file(self, file_name: str) -> None:
+        # TODO: would be nice if walk could skip this updater
+        if file_name == self._view.file_name():
+            self._relevant_file = True
+
+    def diagnostic(self, diagnostic: Diagnostic) -> None:
+        if self._relevant_file:
+            if diagnostic.severity <= settings.show_diagnostics_severity_level:
+                self._regions.setdefault(diagnostic.severity, []).append(range_to_region(diagnostic.range, self._view))
+
+    def end_file(self, file_name: str) -> None:
+        self._relevant_file = False
+
+    def end(self) -> None:
+        for severity in range(DiagnosticSeverity.Error, settings.show_diagnostics_severity_level):
+            region_name = "lsp_" + format_severity(severity)
+            if severity in self._regions:
+                regions = self._regions[severity]
+                scope_name = diagnostic_severity_scopes[severity]
+                self._view.add_regions(
+                    region_name, regions, scope_name, settings.diagnostics_gutter_marker,
+                    UNDERLINE_FLAGS if settings.diagnostics_highlight_style == "underline" else BOX_FLAGS)
+            else:
+                self._view.erase_regions(region_name)
+
+
+class DiagnosticsWalker(object):
+    """ Iterate over diagnostics structure"""
+
+    def __init__(self, subs: 'List[DiagnosticsUpdateWalk]') -> None:
+        self._subscribers = subs
+
+    def walk(self, diagnostics_by_file: 'Dict[str, Dict[str, List[Diagnostic]]]') -> None:
+        self.invoke_each(lambda w: w.begin())
+
+        if diagnostics_by_file is not None:
+            if diagnostics_by_file:
```

Comment:
Why the double if for `diagnostics_by_file`? :-)

Reply:
Good catch, should have left a TODO there. The old `update_diagnostics_panel` had both checks and did nothing at all if `diagnostics_by_file` was `None` - I think that case can't happen anymore, so I'll remove the check!